### PR TITLE
Simplify the implementation of Elim

### PR DESCRIPTION
**Pull request by ppedrot**
Follow-up of #12943. We enforce a few static invariants in the code of Elim.

### CODE REVIEW

**Comment by herbelin** [comment]
Maybe would it lead you too far from what you just want to do now, but I wonder if if could be a right time to specify in a higher-level language what we expect from the ideal `inversion` tactic.

Easier, what about wiring `Indrec.lookup_eliminator` on the registration mechanism?

**Comment by ppedrot** [comment]
> Easier, what about wiring Indrec.lookup_eliminator on the registration mechanism?

This would be a feature PR, while this is just a cleanup PR. In particular, the former would need a bit of design.

**Comment by ppedrot** [comment]
(My secret goal is to get rid of the clenv mechanism in Elim, regardless of the reliance on the old / new unification status.)

**Comment by ppedrot** [comment]
CI green despite the GitHub UI.

File: tactics/elim.ml [hunk 1]
```diff
@@ -202,14 +190,10 @@ let general_decompose recognizer c =
   end
 
 let head_in indl t gl =
-  let env = Proofview.Goal.env gl in
   let sigma = Tacmach.New.project gl in
   try
-    let ity,_ =
-      if !up_to_delta
-      then find_mrectype env sigma t
-      else extract_mrectype sigma t
```
Comment (herbelin) [inline tactics/elim.ml:211]: So, you renounce to explore the "better" up-to-normalization strategy? Maybe compensate this by adding a comment on the code to tell that we would expect a more natural user experience by reasoning up-to-normalization but that this is questionable for compatibility?

**Review by herbelin — APPROVED** [review]
This is the kind of diff for which ticking the `Hide whitespace changes` is necessary! Basically, you are inlining the `mk_elim` combinator, right? Good to me. Assigning.

File: tactics/elim.ml (hunk 1 above)
Comment (ppedrot) [reply tactics/elim.ml:211]: This is dead code that is not reachable by any means, and if ever we wanted to tweak the behaviour it'd be trivial codewise (although probably not compatibility-wise).  Thus I don't really see the interest to keep it around.

Comment (ppedrot) [reply tactics/elim.ml:211]: Actually, this is only used by the `decompose` tactic. I don't really understand why we still have such a tactic in 2020 when it's easy to replace with an Ltac snippet. I'd be curious how much these tactics are used.

Comment (herbelin) [reply tactics/elim.ml:211]: It seems to be used in VST, hoare-tut, exact-real-arithmetic, GeoCoq, CompCert, color, ... Also in the stdlib.

**Pull request merged**